doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,100 +820,189 @@ tactic
-   by substituting :n:`v__i` to each occurrence of :n:`@ident__i`, for
-   i = 1, ..., n. There are no dependencies between the :n:`@ltac_expr__i` and the
-   :n:`@ident__i`.
+   Evaluates each :n:`@let_clause`, substitutes the bound variables into
+   :n:`@ltac_expr` and then evaluates :n:`@ltac_expr`.  There are no dependencies
+   between the :n:`@let_clause`\s.
 
-   Local definitions can be made recursive by using :n:`let rec` instead of :n:`let`.
-   In this latter case, the definitions are evaluated lazily so that the rec
-   keyword can be used also in non-recursive cases so as to avoid the eager
+   Creates a local definition.
+   Using :tacn:`let` `rec` makes the definition recursive and causes
+   the definitions to be evaluated lazily.  Thus, :tacn:`let` `rec`
+   can be used in non-recursive cases to avoid eager
    evaluation of local definitions.
 
    .. but rec changes the binding!!
 
+   :tacn:`let` is a :token:`binder_tactic`.
+
 Application
 ~~~~~~~~~~~
 
 An application is an expression of the following form:
 
-.. tacn:: @qualid {+ @tacarg}
+.. tacn:: @qualid {* @tactic_arg }
 
    The reference :n:`@qualid` must be bound to some defined tactic definition
-   expecting at least as many arguments as the provided :n:`tacarg`. The
-   expressions :n:`@ltac_expr__i` are evaluated to :n:`v__i`, for i = 1, ..., n.
+   expecting at least as many arguments as the provided :n:`@tactic_arg`. The
+   expressions :n:`@tactic_arg__i` are evaluated to :n:`v__i`, for i = 1, ..., n.
 
    .. what expressions ??
 
-Function construction
-~~~~~~~~~~~~~~~~~~~~~
+Function construction: fun
+~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 A parameterized tactic can be built anonymously (without resorting to
 local definitions) with:
 
-.. tacn:: fun {+ @ident} => @ltac_expr
+.. tacn:: fun {+ @name } => @ltac_expr
 
    Indeed, local definitions of functions are a syntactic sugar for binding
    a :n:`fun` tactic to an identifier.
 
-Pattern matching on terms
-~~~~~~~~~~~~~~~~~~~~~~~~~
+   :tacn:`fun` is a :token:`binder_tactic`.
+
+Pattern matching on terms: match
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tacn:: @match_key @ltac_expr with {? %| } {+| @match_pattern => @ltac_expr } end
+   :name: match; lazymatch; multimatch
+
+   .. insertprodn match_key cpattern
+
+   .. prodn::
+      match_key ::= match
+      | multimatch
+      | lazymatch
+      match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+      cpattern ::= @term
+
+   .. todo I don't fully understand these descriptions.  I think some concepts
+      such as backtracking need to be explained first and that the lanugage
+      can probably be simplified.
+
+   :tacn:`match`, :tacn:`lazymatch` and :tacn:`multimatch` are :token:`ltac_expr1`\s.
+
+   `match`
+      TBA
+
+     .. todo PR: How does match differ from multimatch and lazymatch?  Does it consider
+        multiple matches if tactics fail, but won't backtrack if tactics after the match
+        fail?
```

Comment:
This can indeed be inferred from the description of `multimatch` and `lazymatch`. What you wrote is precisely it AFAICT. This is also confirmed by the comment about `match` being expressed in terms of `once` and `multimatch`.

Reply:
Any idea why this doesn't work?  I wanted to verify that match uses nonlinear unification.

```
match (add 1 1) with
| (add ?X ?X) => idtac "yes"
| _ => idtac "no"
end.

(* The reference add was not found in the current environment. *)
```

Reply:
Use `constr:(add 1 1)`.  The default interpretation is to treat application as Ltac function application, not term application.

Reply:
No change.  The error with either form is on the second `add`.

```
match constr:(add 1 1) with
| (add ?X ?X) => idtac "yes"
| _ => idtac "no"
end.

(* The reference add was not found in the current environment. *)
```

Reply:
Ah, that's because it's called `plus`, not `add`.

Reply:
Right, thanks.